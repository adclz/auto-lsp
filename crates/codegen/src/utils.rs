/*
This file is part of auto-lsp.
Copyright (C) 2025 CLAUZEL Adrien

auto-lsp is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>
*/

pub(crate) fn sanitize_string(string: &str) -> String {
    if let Some(v) = PUNCTUATION.get(string) {
        v.to_string()
    } else if let Some(v) = RUST_KEYWORDS.get(string) {
        v.to_string()
    } else {
        let mut result = String::new();
        for c in string.chars() {
            if c == '_' {
                result.push(c);
            } else if let Some(v) = PUNCTUATION.get(&c.to_string()) {
                result.push_str(v);
            } else {
                result.push(c);
            }
        }
        result.replace(" ", "_")
    }
}

pub(crate) fn sanitize_string_to_pascal(string: &str) -> String {
    if let Some(v) = PUNCTUATION.get(string) {
        v.to_string()
    } else if let Some(v) = RUST_KEYWORDS.get(string) {
        v.to_string()
    } else {
        let mut result = String::new();
        for c in string.chars() {
            if c == '_' {
                result.push(c);
            } else if let Some(v) = PUNCTUATION.get(&c.to_string()) {
                result.push_str(v);
            } else {
                result.push(c);
            }
        }
        result
            .replace(" ", "_")
            .split('_')
            .map(|word| {
                let mut chars = word.chars();
                match chars.next() {
                    Some(first) => format!(
                        "{}{}",
                        first.to_ascii_uppercase(),
                        chars.collect::<String>()
                    ),
                    None => String::new(),
                }
            })
            .collect()
    }
}

pub static PUNCTUATION: phf::Map<&'static str, &'static str> = phf::phf_map! {
    "{" => "LeftCurly",
    "}" => "RightCurly",
    "(" => "LeftParen",
    ")" => "RightParen",
    "[" => "LeftBracket",
    "]" => "RightBracket",
    "," => "Comma",
    ":" => "Colon",
    ";" => "Semicolon",
    "." => "Dot",
    "'" => "Quote",
    "\"" => "DoubleQuote",
    "@" => "At",
    "!" => "Bang",
    "#" => "Hash",
    "$" => "Dollar",
    "%" => "Percent",
    "^" => "Caret",
    "&" => "Ampersand",
    "*" => "Star",
    "-" => "Minus",
    "_" => "Underscore",
    "+" => "Plus",
    "=" => "Equal",
    ">" => "Greater",
    "<" => "Less",
    "|" => "Pipe",
    "~" => "Tilde",
    "/" => "Slash",
    "\\" => "Backslash",
    "//" => "SlashSlash",
    "//=>" => "SlashSlashEqual",
    "//=" => "SlashSlashEqual",
    "/=" => "SlashEqual",
    "/>" => "SlashGreater",
    "/?" => "SlashQuestion",
    "/??" => "SlashNullish",
    "/*" => "SlashStar",
    "*/" => "StarSlash",
    "+++" => "PlusPlusPlus",
    "!!" => "BangBang",
    "!!=" => "BangBangEqual",
    "!!?" => "BangBangQuestion",
    "!!??" => "BangBangNullish",
    "!!???" => "BangBangNullishQuestion",
    "?" => "Question",
    "->" => "Arrow",
    "=>" => "FatArrow",
    "++" => "PlusPlus",
    "--" => "MinusMinus",
    "&&" => "And",
    "||" => "Or",
    "==" => "EqualEqual",
    "!=" => "NotEqual",
    ">=" => "GreaterEqual",
    "<=" => "LessEqual",
    "===" => "TripleEqual",
    "!==" => "NotTripleEqual",
    "<<" => "ShiftLeft",
    ">>" => "ShiftRight",
    ">>>" => "ShiftRightUnsigned",
    "+=" => "PlusEqual",
    "-=" => "MinusEqual",
    "*=" => "StarEqual",
    "%=" => "PercentEqual",
    "&=" => "AmpersandEqual",
    "|=" => "PipeEqual",
    "^=" => "CaretEqual",
    "&&=" => "AndEqual",
    "||=" => "OrEqual",
    "??=" => "NullishEqual",
    "??" => "Nullish",
    "???" => "NullishQuestion",
    "**" => "StarStar",
    "**=" => "StarStarEqual",
    "<>" => "LessGreater",
    "<=>" => "LessGreaterEqual",
    "<!" => "LessBang",
    "</" => "LessSlash",

};

pub static RUST_KEYWORDS: phf::Map<&'static str, &'static str> = phf::phf_map! {
    "abstract" => "Abstract",
    "as" => "As",
    "async" => "Async",
    "await" => "Await",
    "break" => "Break",
    "const" => "Const",
    "continue" => "Continue",
    "crate" => "Crate",
    "dyn" => "Dyn",
    "else" => "Else",
    "enum" => "Enum",
    "extern" => "Extern",
    "false" => "False",
    "final" => "Final",
    "fn" => "Fn",
    "for" => "For",
    "if" => "If",
    "impl" => "Impl",
    "in" => "In",
    "let" => "Let",
    "loop" => "Loop",
    "match" => "Match",
    "mod" => "Mod",
    "move" => "Move",
    "mut" => "Mut",
    "pub" => "Pub",
    "ref" => "Ref",
    "return" => "Return",
    "self" => "Self",
    "static" => "Static",
    "struct" => "Struct",
    "super" => "Super",
    "trait" => "Trait",
    "true" => "True",
    "type" => "Type",
    "unsafe" => "Unsafe",
    "use" => "Use",
    "where" => "Where",
    "while" => "While",
    "with" => "With",
    "yield" => "Yield",
    "None" => "_None",
    "Some" => "_Some",
    "Ok" => "_Ok",
    "Err" => "_Err",
    "Result" => "_Result",
    "Option" => "_Option",
    "Vec" => "_Vec",
};
