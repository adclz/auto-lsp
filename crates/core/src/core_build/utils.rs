/*
This file is part of auto-lsp.
Copyright (C) 2025 CLAUZEL Adrien

auto-lsp is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>
*/

pub(crate) fn intersecting_ranges(
    range1: &std::ops::Range<usize>,
    range2: &tree_sitter::Range,
) -> bool {
    range1.start <= range2.start_byte && range1.end >= range2.end_byte
}

pub(crate) fn tree_sitter_range_to_lsp_range(range: &tree_sitter::Range) -> lsp_types::Range {
    let start = range.start_point;
    let end = range.end_point;
    lsp_types::Range {
        start: lsp_types::Position {
            line: start.row as u32,
            character: start.column as u32,
        },
        end: lsp_types::Position {
            line: end.row as u32,
            character: end.column as u32,
        },
    }
}
