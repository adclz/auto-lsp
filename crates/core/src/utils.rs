#[macro_export]
macro_rules! dispatch_once {
    ($node:expr, [$($ty:ty => $method:ident($($param:expr),*)),*]) => {
        $(
            if let Some(n) = $node.downcast_ref::<$ty>() {
                return n.$method($($param),*);
            };
        )*
    };
}

#[macro_export]
macro_rules! dispatch {
    ($node:expr, [$($ty:ty => $method:ident($($param:expr),*)),*]) => {
        $(
            if let Some(n) = $node.downcast_ref::<$ty>() {
                n.$method($($param),*)?;
            };
        )*
    };
}