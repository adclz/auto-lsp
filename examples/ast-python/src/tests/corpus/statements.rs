/*
This file is part of auto-lsp.
Copyright (C) 2025 CLAUZEL Adrien

auto-lsp is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>
*/

use super::utils::Result;
use crate::snap;

#[test]
fn import_statements() -> Result {
    snap!(
        r#"
import a, b
import b.c as d
import a.b.c
"#
    )
}

#[test]
fn import_from_statements() -> Result {
    snap!(
        r#"
from a import b
from a import *
from a import (b, c)
from a.b import c
from . import b
from .. import b
from .a import b
from ..a import b
"#
    )
}

#[test]
fn future_import_statements() -> Result {
    snap!(
        r#"
from __future__ import print_statement
from __future__ import python4
from __future__ import (absolute_import, division, print_function,
                        unicode_literals)
"#
    )
}

#[test]
fn print_statements() -> Result {
    snap!(
        r#"
print a
print b, c
print 0 or 1, 1 or 0,
print 0 or 1
print not True
"#
    )
}

#[test]
fn print_statements_with_redirection() -> Result {
    snap!(
        r#"
print >> a
print >> a, "b", "c"
"#
    )
}

#[test]
fn assert_statements() -> Result {
    snap!(
        r#"
assert a
assert b, c
"#
    )
}

#[test]
fn expression_statements() -> Result {
    snap!(
        r#"
a
b + c
1, 2, 3
1, 2, 3,
"#
    )
}

#[test]
fn delete_statements() -> Result {
    snap!(
        r#"
del a[1], b[2]
"#
    )
}

#[test]
fn control_flow_statements() -> Result {
    snap!(
        r#"
while true:
  pass
  break
  continue
"#
    )
}

#[test]
fn return_statements() -> Result {
    snap!(
        r#"
return
return a + b, c
return not b
"#
    )
}

#[test]
fn if_statements() -> Result {
    snap!(
        r#"
if a:
  b
  c
"#
    )
}

#[test]
fn if_else_statements() -> Result {
    snap!(
        r#"
if a:
  b
elif c:
  d
else:
  f

if a:
  b
else:
  f

if a: b

if a: b; c
"#
    )
}

#[test]
fn nested_if_statements() -> Result {
    snap!(
        r#"
if a:
  if b:
    c
  else:
    if e:
      f
g
"#
    )
}

#[test]
fn while_statements() -> Result {
    snap!(
        r#"
while a:
  b

while c:
  d
else:
  e
  f
"#
    )
}

#[test]
fn for_statements() -> Result {
    snap!(
        r#"
for line, i in lines:
  print line
  for character, j in line:
    print character
else:
  print x

for x, in [(1,), (2,), (3,)]:
  x
"#
    )
}

#[test]
fn try_statements() -> Result {
    snap!(
        r#"
try:
  a
except b:
  c
except d as e:
  f
except g, h:
  i
except:
  j

try:
  a
except b:
  c
  d
else:
  e
finally:
  f

try:
  a
except* b:
  c
except* d as e:
  f
else:
  g
finally:
  h
"#
    )
}

#[test]
fn with_statements() -> Result {
    snap!(
        r#"
with a as b:
  c

with (open('d') as d,
      open('e') as e):
  f

with e as f, g as h,:
  i
"#
    )
}

#[test]
fn async_function_definitions() -> Result {
    snap!(
        r#"
async def a():
  b

async def c(d):
  e

async def g(g, h,):
  i

async def c(a: str):
  a

async def c(a: b.c):
  a

async def d(a: Sequence[T]) -> T:
  a

async def i(a, b=c, *c, **d):
  a

async def d(a: str) -> None:
  return None

async def d(a:str="default", b=c) -> None:
  return None
"#
    )
}

#[test]
fn function_definitions() -> Result {
    snap!(
        r#"
def e((a,b)):
  return (a,b)

def e(*list: str):
  pass

def e(**list: str):
  pass

def f():
  nonlocal a

def g(h, i, /, j, *, k=100, **kwarg):
  return h,i,j,k,kwarg

def h(*a):
  i((*a))
  j(((*a)))

def foo():
    pass \
\
\

"#
    )
}

#[test]
fn empty_blocks() -> Result {
    snap!(
        r#"
# These are not actually valid python; blocks
# must contain at least one statement. But we
# allow them because error recovery for empty
# blocks doesn't work very well otherwise.
def a(b, c):

if d:
  print e
  while f():
"#
    )
}

#[test]
fn class_definitions() -> Result {
    snap!(
        r#"
class A:
  def b(self):
    return c
class B():
  pass
class B(method1):
  def method1(self):
    return
class C(method1, Sequence[T]):
  pass
class D(Sequence[T, U]):
  pass
"#
    )
}
