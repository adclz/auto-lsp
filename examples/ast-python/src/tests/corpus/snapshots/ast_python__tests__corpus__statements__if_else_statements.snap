---
source: src/tests/corpus/statements.rs
expression: module
---
Module {
    children: [
        CompoundStatement(
            IfStatement(
                IfStatement {
                    alternative: [
                        ElifClause(
                            ElifClause {
                                consequence: Block {
                                    alternative: [],
                                    children: [
                                        SimpleStatement(
                                            ExpressionStatement(
                                                ExpressionStatement {
                                                    children: [
                                                        Expression(
                                                            PrimaryExpression(
                                                                Identifier(
                                                                    Identifier {
                                                                        [RANGE]
                                                                        _id: 10,
                                                                        _parent: Some(
                                                                            9,
                                                                        ),
                                                                    },
                                                                ),
                                                            ),
                                                        ),
                                                    ],
                                                    [RANGE]
                                                    _id: 9,
                                                    _parent: Some(
                                                        8,
                                                    ),
                                                },
                                            ),
                                        ),
                                    ],
                                    [RANGE]
                                    _id: 8,
                                    _parent: Some(
                                        6,
                                    ),
                                },
                                condition: PrimaryExpression(
                                    Identifier(
                                        Identifier {
                                            [RANGE]
                                            _id: 7,
                                            _parent: Some(
                                                6,
                                            ),
                                        },
                                    ),
                                ),
                                [RANGE]
                                _id: 6,
                                _parent: Some(
                                    1,
                                ),
                            },
                        ),
                        ElseClause(
                            ElseClause {
                                body: Block {
                                    alternative: [],
                                    children: [
                                        SimpleStatement(
                                            ExpressionStatement(
                                                ExpressionStatement {
                                                    children: [
                                                        Expression(
                                                            PrimaryExpression(
                                                                Identifier(
                                                                    Identifier {
                                                                        [RANGE]
                                                                        _id: 14,
                                                                        _parent: Some(
                                                                            13,
                                                                        ),
                                                                    },
                                                                ),
                                                            ),
                                                        ),
                                                    ],
                                                    [RANGE]
                                                    _id: 13,
                                                    _parent: Some(
                                                        12,
                                                    ),
                                                },
                                            ),
                                        ),
                                    ],
                                    [RANGE]
                                    _id: 12,
                                    _parent: Some(
                                        11,
                                    ),
                                },
                                [RANGE]
                                _id: 11,
                                _parent: Some(
                                    1,
                                ),
                            },
                        ),
                    ],
                    consequence: Block {
                        alternative: [],
                        children: [
                            SimpleStatement(
                                ExpressionStatement(
                                    ExpressionStatement {
                                        children: [
                                            Expression(
                                                PrimaryExpression(
                                                    Identifier(
                                                        Identifier {
                                                            [RANGE]
                                                            _id: 5,
                                                            _parent: Some(
                                                                4,
                                                            ),
                                                        },
                                                    ),
                                                ),
                                            ),
                                        ],
                                        [RANGE]
                                        _id: 4,
                                        _parent: Some(
                                            3,
                                        ),
                                    },
                                ),
                            ),
                        ],
                        [RANGE]
                        _id: 3,
                        _parent: Some(
                            1,
                        ),
                    },
                    condition: PrimaryExpression(
                        Identifier(
                            Identifier {
                                [RANGE]
                                _id: 2,
                                _parent: Some(
                                    1,
                                ),
                            },
                        ),
                    ),
                    [RANGE]
                    _id: 1,
                    _parent: Some(
                        0,
                    ),
                },
            ),
        ),
        CompoundStatement(
            IfStatement(
                IfStatement {
                    alternative: [
                        ElseClause(
                            ElseClause {
                                body: Block {
                                    alternative: [],
                                    children: [
                                        SimpleStatement(
                                            ExpressionStatement(
                                                ExpressionStatement {
                                                    children: [
                                                        Expression(
                                                            PrimaryExpression(
                                                                Identifier(
                                                                    Identifier {
                                                                        [RANGE]
                                                                        _id: 23,
                                                                        _parent: Some(
                                                                            22,
                                                                        ),
                                                                    },
                                                                ),
                                                            ),
                                                        ),
                                                    ],
                                                    [RANGE]
                                                    _id: 22,
                                                    _parent: Some(
                                                        21,
                                                    ),
                                                },
                                            ),
                                        ),
                                    ],
                                    [RANGE]
                                    _id: 21,
                                    _parent: Some(
                                        20,
                                    ),
                                },
                                [RANGE]
                                _id: 20,
                                _parent: Some(
                                    15,
                                ),
                            },
                        ),
                    ],
                    consequence: Block {
                        alternative: [],
                        children: [
                            SimpleStatement(
                                ExpressionStatement(
                                    ExpressionStatement {
                                        children: [
                                            Expression(
                                                PrimaryExpression(
                                                    Identifier(
                                                        Identifier {
                                                            [RANGE]
                                                            _id: 19,
                                                            _parent: Some(
                                                                18,
                                                            ),
                                                        },
                                                    ),
                                                ),
                                            ),
                                        ],
                                        [RANGE]
                                        _id: 18,
                                        _parent: Some(
                                            17,
                                        ),
                                    },
                                ),
                            ),
                        ],
                        [RANGE]
                        _id: 17,
                        _parent: Some(
                            15,
                        ),
                    },
                    condition: PrimaryExpression(
                        Identifier(
                            Identifier {
                                [RANGE]
                                _id: 16,
                                _parent: Some(
                                    15,
                                ),
                            },
                        ),
                    ),
                    [RANGE]
                    _id: 15,
                    _parent: Some(
                        0,
                    ),
                },
            ),
        ),
        CompoundStatement(
            IfStatement(
                IfStatement {
                    alternative: [],
                    consequence: Block {
                        alternative: [],
                        children: [
                            SimpleStatement(
                                ExpressionStatement(
                                    ExpressionStatement {
                                        children: [
                                            Expression(
                                                PrimaryExpression(
                                                    Identifier(
                                                        Identifier {
                                                            [RANGE]
                                                            _id: 28,
                                                            _parent: Some(
                                                                27,
                                                            ),
                                                        },
                                                    ),
                                                ),
                                            ),
                                        ],
                                        [RANGE]
                                        _id: 27,
                                        _parent: Some(
                                            26,
                                        ),
                                    },
                                ),
                            ),
                        ],
                        [RANGE]
                        _id: 26,
                        _parent: Some(
                            24,
                        ),
                    },
                    condition: PrimaryExpression(
                        Identifier(
                            Identifier {
                                [RANGE]
                                _id: 25,
                                _parent: Some(
                                    24,
                                ),
                            },
                        ),
                    ),
                    [RANGE]
                    _id: 24,
                    _parent: Some(
                        0,
                    ),
                },
            ),
        ),
        CompoundStatement(
            IfStatement(
                IfStatement {
                    alternative: [],
                    consequence: Block {
                        alternative: [],
                        children: [
                            SimpleStatement(
                                ExpressionStatement(
                                    ExpressionStatement {
                                        children: [
                                            Expression(
                                                PrimaryExpression(
                                                    Identifier(
                                                        Identifier {
                                                            [RANGE]
                                                            _id: 33,
                                                            _parent: Some(
                                                                32,
                                                            ),
                                                        },
                                                    ),
                                                ),
                                            ),
                                        ],
                                        [RANGE]
                                        _id: 32,
                                        _parent: Some(
                                            31,
                                        ),
                                    },
                                ),
                            ),
                            SimpleStatement(
                                ExpressionStatement(
                                    ExpressionStatement {
                                        children: [
                                            Expression(
                                                PrimaryExpression(
                                                    Identifier(
                                                        Identifier {
                                                            [RANGE]
                                                            _id: 35,
                                                            _parent: Some(
                                                                34,
                                                            ),
                                                        },
                                                    ),
                                                ),
                                            ),
                                        ],
                                        [RANGE]
                                        _id: 34,
                                        _parent: Some(
                                            31,
                                        ),
                                    },
                                ),
                            ),
                        ],
                        [RANGE]
                        _id: 31,
                        _parent: Some(
                            29,
                        ),
                    },
                    condition: PrimaryExpression(
                        Identifier(
                            Identifier {
                                [RANGE]
                                _id: 30,
                                _parent: Some(
                                    29,
                                ),
                            },
                        ),
                    ),
                    [RANGE]
                    _id: 29,
                    _parent: Some(
                        0,
                    ),
                },
            ),
        ),
    ],
    [RANGE]
    _id: 0,
    _parent: None,
}
